File: packages/modules/lodash/core/kotlin/src/test/kotlin/lodash/StringTest.kt
package lodash.test

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class StringTest {

    @Test
    fun camelCase() {
        assertEquals("fooBar", lodash.camelCase("Foo Bar"))
        assertEquals("fooBar", lodash.camelCase("--foo-bar--"))
        assertEquals("fooBar", lodash.camelCase("foo_bar"))
        assertEquals("fooBar", lodash.camelCase("FOO BAR"))
        assertEquals("", lodash.camelCase(null))
    }

    @Test
    fun capitalize() {
        assertEquals("Hello", lodash.capitalize("hello"))
        assertEquals("Hello", lodash.capitalize("HELLO"))
        assertEquals("Hello world", lodash.capitalize("hello world"))
        assertEquals("", lodash.capitalize(null))
    }

    @Test
    fun endsWith() {
        assertTrue(lodash.endsWith("hello", "lo"))
        assertFalse(lodash.endsWith("hello", "el"))
        assertFalse(lodash.endsWith(null, "lo"))
        assertFalse(lodash.endsWith("hello", null))
        assertTrue(lodash.endsWith("hello", "hel", 3))
        assertFalse(lodash.endsWith("hello", "lo", -1))
        assertFalse(lodash.endsWith("hello", "lo", 0))
    }

    @Test
    fun escape() {
        assertEquals("&amp;", lodash.escape("&"))
        assertEquals("&lt;&gt;", lodash.escape("<>"))
        assertEquals("&quot;hello&quot;", lodash.escape("\"hello\""))
        assertEquals("&#39;", lodash.escape("'"))
        assertEquals("", lodash.escape(null))
    }

    @Test
    fun kebabCase() {
        assertEquals("foo-bar", lodash.kebabCase("Foo Bar"))
        assertEquals("foobar", lodash.kebabCase("fooBar"))
        assertEquals("foo-bar", lodash.kebabCase("__FOO_BAR__"))
        assertEquals("", lodash.kebabCase(null))
    }

    @Test
    fun lowerFirst() {
        assertEquals("hello", lodash.lowerFirst("Hello"))
        assertEquals("hello", lodash.lowerFirst("hello"))
        assertEquals("hELLO", lodash.lowerFirst("HELLO"))
        assertEquals("", lodash.lowerFirst(null))
    }

    @Test
    fun pad() {
        assertEquals("  hi  ", lodash.pad("hi", 6))
        assertEquals("--hi--", lodash.pad("hi", 6, "-"))
        assertEquals("hi", lodash.pad("hi", 2))
        assertEquals("hi", lodash.pad("hi", 1))
        assertEquals("", lodash.pad(null, 6))
    }

    @Test
    fun padEnd() {
        assertEquals("hi    ", lodash.padEnd("hi", 6))
        assertEquals("hi----", lodash.padEnd("hi", 6, "-"))
        assertEquals("hi", lodash.padEnd("hi", 2))
        assertEquals("", lodash.padEnd(null, 6))
    }

    @Test
    fun padStart() {
        assertEquals("    hi", lodash.padStart("hi", 6))
        assertEquals("----hi", lodash.padStart("hi", 6, "-"))
        assertEquals("hi", lodash.padStart("hi", 2))
        assertEquals("", lodash.padStart(null, 6))
    }

    @Test
    fun repeatStr() {
        assertEquals("aaa", lodash.repeat("a", 3))
        assertEquals("", lodash.repeat("a", 0))
        assertEquals("", lodash.repeat("a", -1))
        assertEquals("", lodash.repeat(null, 3))
    }

    @Test
    fun replace() {
        assertEquals("hxllo", lodash.replace("hello", "el", "xl"))
        assertEquals("hello", lodash.replace("hello", "xyz", "abc"))
        assertEquals("", lodash.replace(null, "el", "xl"))
    }

    @Test
    fun snakeCase() {
        assertEquals("foo_bar", lodash.snakeCase("Foo Bar"))
        assertEquals("foobar", lodash.snakeCase("fooBar"))
        assertEquals("foo_bar", lodash.snakeCase("--FOO-BAR--"))
        assertEquals("", lodash.snakeCase(null))
    }

    @Test
    fun split() {
        assertEquals(listOf("a", "b", "c"), lodash.split("a,b,c", ","))
        assertEquals(listOf("a", "b"), lodash.split("a,b,c", ",", 2))
        assertEquals(listOf("a,b,c"), lodash.split("a,b,c", null))
        assertEquals(emptyList<String>(), lodash.split(null, ","))
    }

    @Test
    fun startCase() {
        assertEquals("Foobar", lodash.startCase("fooBar"))
        assertEquals("Foo Bar", lodash.startCase("FOO BAR"))
        assertEquals("Foo Bar", lodash.startCase("--foo-bar--"))
        assertEquals("", lodash.startCase(null))
    }

    @Test
    fun startsWith() {
        assertTrue(lodash.startsWith("hello", "hel"))
        assertFalse(lodash.startsWith("hello", "lo"))
        assertTrue(lodash.startsWith("hello", "hel", 0))
        assertTrue(lodash.startsWith("hello", "lo", 3))
        assertFalse(lodash.startsWith(null, "hel"))
        assertFalse(lodash.startsWith("hello", null))
    }

    @Test
    fun toLower() {
        assertEquals("hello", lodash.toLower("HELLO"))
        assertEquals("hello", lodash.toLower("hello"))
        assertEquals("", lodash.toLower(null))
    }

    @Test
    fun toUpper() {
        assertEquals("HELLO", lodash.toUpper("hello"))
        assertEquals("HELLO", lodash.toUpper("HELLO"))
        assertEquals("", lodash.toUpper(null))
    }

    @Test
    fun truncate() {
        assertEquals("hello...", lodash.truncate("hello world", 8))
        assertEquals("hello world", lodash.truncate("hello world", 20))
        assertEquals("hello!!!", lodash.truncate("hello world", 8, "!!!"))
        assertEquals("", lodash.truncate(null))
    }

    @Test
    fun unescape() {
        assertEquals("&", lodash.unescape("&amp;"))
        assertEquals("<>", lodash.unescape("&lt;&gt;"))
        assertEquals("\"hello\"", lodash.unescape("&quot;hello&quot;"))
        assertEquals("'", lodash.unescape("&#39;"))
        assertEquals("", lodash.unescape(null))
    }

    @Test
    fun upperFirst() {
        assertEquals("Hello", lodash.upperFirst("hello"))
        assertEquals("Hello", lodash.upperFirst("Hello"))
        assertEquals("HELLO", lodash.upperFirst("hELLO"))
        assertEquals("", lodash.upperFirst(null))
    }

    @Test
    fun words() {
        assertEquals(listOf("hello", "world"), lodash.words("hello world"))
        assertEquals(listOf("abc123"), lodash.words("abc123"))
        assertEquals(emptyList<String>(), lodash.words(null))
        assertEquals(emptyList<String>(), lodash.words("!@#"))
    }
}

File: packages/modules/lodash/core/go/core/string/string_test.go
package string

import (
	"testing"
)

func TestCamelCase(t *testing.T) {
	if got := CamelCase("Foo Bar"); got != "fooBar" {
		t.Errorf("expected 'fooBar', got '%s'", got)
	}
	if got := CamelCase("--foo-bar--"); got != "fooBar" {
		t.Errorf("expected 'fooBar', got '%s'", got)
	}
	if got := CamelCase("FOO BAR"); got != "fooBar" {
		t.Errorf("expected 'fooBar', got '%s'", got)
	}
}

func TestCapitalize(t *testing.T) {
	if got := Capitalize("FRED"); got != "Fred" {
		t.Errorf("expected 'Fred', got '%s'", got)
	}
	if got := Capitalize(""); got != "" {
		t.Errorf("expected '', got '%s'", got)
	}
	if got := Capitalize("hello"); got != "Hello" {
		t.Errorf("expected 'Hello', got '%s'", got)
	}
}

func TestDeburr(t *testing.T) {
	if got := Deburr("déjà vu"); got != "deja vu" {
		t.Errorf("expected 'deja vu', got '%s'", got)
	}
	if got := Deburr("über"); got != "uber" {
		t.Errorf("expected 'uber', got '%s'", got)
	}
	if got := Deburr("abc"); got != "abc" {
		t.Errorf("expected 'abc', got '%s'", got)
	}
}

func TestEndsWith(t *testing.T) {
	if EndsWith("abc", 'c', 0) {
		t.Errorf("expected false for position 0")
	}
	if !EndsWith("abc", 'c', 3) {
		t.Errorf("expected true for full string")
	}
	if EndsWith("abc", 'b', 1) {
		t.Errorf("expected false (first char 'a')")
	}
	if !EndsWith("abc", 'b', 2) {
		t.Errorf("expected true (first 2 chars 'ab')")
	}
}

func TestEscape(t *testing.T) {
	if got := Escape("a&b<c>d\"e'"); got != "a&amp;b&lt;c&gt;d&quot;e&#39;" {
		t.Errorf("expected escaped string, got '%s'", got)
	}
	if got := Escape("hello"); got != "hello" {
		t.Errorf("expected unchanged, got '%s'", got)
	}
}

func TestEscapeRegExp(t *testing.T) {
	if got := EscapeRegExp("hello.world"); got != "hello\\.world" {
		t.Errorf("expected 'hello\\.world', got '%s'", got)
	}
}

func TestKebabCase(t *testing.T) {
	if got := KebabCase("Foo Bar"); got != "foo-bar" {
		t.Errorf("expected 'foo-bar', got '%s'", got)
	}
	if got := KebabCase("fooBar"); got != "foo-bar" {
		t.Errorf("expected 'foo-bar', got '%s'", got)
	}
}

func TestLowerCase(t *testing.T) {
	if got := LowerCase("--Foo-Bar--"); got != "foo bar" {
		t.Errorf("expected 'foo bar', got '%s'", got)
	}
	if got := LowerCase("fooBar"); got != "foo bar" {
		t.Errorf("expected 'foo bar', got '%s'", got)
	}
}

func TestLowerFirst(t *testing.T) {
	if got := LowerFirst("Fred"); got != "fred" {
		t.Errorf("expected 'fred', got '%s'", got)
	}
	if got := LowerFirst("FRED"); got != "fRED" {
		t.Errorf("expected 'fRED', got '%s'", got)
	}
	if got := LowerFirst(""); got != "" {
		t.Errorf("expected '', got '%s'", got)
	}
}

func TestPad(t *testing.T) {
	if got := Pad("abc", 7, "_-"); got != "_-abc_-" {
		t.Errorf("expected '_-abc_-', got '%s'", got)
	}
	if got := Pad("abc", 3, "*"); got != "abc" {
		t.Errorf("expected 'abc', got '%s'", got)
	}
	if got := Pad("abc", 5, ""); got != " abc " {
		t.Errorf("expected ' abc ', got '%s'", got)
	}
}

func TestPadEnd(t *testing.T) {
	if got := PadEnd("abc", 6, "*"); got != "abc***" {
		t.Errorf("expected 'abc***', got '%s'", got)
	}
	if got := PadEnd("abc", 2, "*"); got != "abc" {
		t.Errorf("expected 'abc', got '%s'", got)
	}
	if got := PadEnd("abc", 5, ""); got != "abc  " {
		t.Errorf("expected 'abc  ', got '%s'", got)
	}
}

func TestPadStart(t *testing.T) {
	if got := PadStart("abc", 6, "*"); got != "***abc" {
		t.Errorf("expected '***abc', got '%s'", got)
	}
	if got := PadStart("abc", 2, "*"); got != "abc" {
		t.Errorf("expected 'abc', got '%s'", got)
	}
	if got := PadStart("abc", 5, ""); got != "  abc" {
		t.Errorf("expected '  abc', got '%s'", got)
	}
}

func TestParseInt(t *testing.T) {
	if got := ParseInt("42", 10); got != 42 {
		t.Errorf("expected 42, got %d", got)
	}
	if got := ParseInt("ff", 16); got != 255 {
		t.Errorf("expected 255, got %d", got)
	}
	if got := ParseInt("invalid", 10); got != 0 {
		t.Errorf("expected 0, got %d", got)
	}
}

func TestRepeat(t *testing.T) {
	if got := Repeat("*", 3); got != "***" {
		t.Errorf("expected '***', got '%s'", got)
	}
	if got := Repeat("abc", 0); got != "" {
		t.Errorf("expected '', got '%s'", got)
	}
}

func TestReplace(t *testing.T) {
	if got := Replace("hello world", "world", "there"); got != "hello there" {
		t.Errorf("expected 'hello there', got '%s'", got)
	}
	if got := Replace("abc", "x", "y"); got != "abc" {
		t.Errorf("expected 'abc', got '%s'", got)
	}
}

func TestSnakeCase(t *testing.T) {
	if got := SnakeCase("Foo Bar"); got != "foo_bar" {
		t.Errorf("expected 'foo_bar', got '%s'", got)
	}
	if got := SnakeCase("fooBar"); got != "foo_bar" {
		t.Errorf("expected 'foo_bar', got '%s'", got)
	}
}

func TestSplit(t *testing.T) {
	result := Split("a-b-c", "-", 2)
	if len(result) != 2 || result[0] != "a" || result[1] != "b-c" {
		t.Errorf("expected [a b-c], got %v", result)
	}

	result2 := Split("a-b-c", "-", 0)
	if len(result2) != 0 {
		t.Errorf("expected [], got %v", result2)
	}

	result3 := Split("a-b-c", "-", -1)
	if len(result3) != 3 {
		t.Errorf("expected [a b c], got %v", result3)
	}
}

func TestStartCase(t *testing.T) {
	if got := StartCase("--foo-bar--"); got != "Foo Bar" {
		t.Errorf("expected 'Foo Bar', got '%s'", got)
	}
	if got := StartCase("FOO BAR"); got != "Foo Bar" {
		t.Errorf("expected 'Foo Bar', got '%s'", got)
	}
}

func TestStartsWith(t *testing.T) {
	if !StartsWith("abc", 'a', 0) {
		t.Errorf("expected true")
	}
	if StartsWith("abc", 'a', 1) {
		t.Errorf("expected false")
	}
	if !StartsWith("abc", 'b', 1) {
		t.Errorf("expected true")
	}
}

func TestTemplate(t *testing.T) {
	data := map[string]string{"name": "world", "greeting": "hello"}
	if got := Template("{{greeting}} {{name}}!", data); got != "hello world!" {
		t.Errorf("expected 'hello world!', got '%s'", got)
	}

	if got := Template("no placeholders", data); got != "no placeholders" {
		t.Errorf("expected 'no placeholders', got '%s'", got)
	}

	if got := Template("{{missing}}", data); got != "" {
		t.Errorf("expected '', got '%s'", got)
	}

	if got := Template("{{unclosed", data); got != "{{unclosed" {
		t.Errorf("expected '{{unclosed', got '%s'", got)
	}
}

func TestToLower(t *testing.T) {
	if got := ToLower("ABC"); got != "abc" {
		t.Errorf("expected 'abc', got '%s'", got)
	}
}

func TestToUpper(t *testing.T) {
	if got := ToUpper("abc"); got != "ABC" {
		t.Errorf("expected 'ABC', got '%s'", got)
	}
}

func TestTruncate(t *testing.T) {
	t.Run("default length 30", func(t *testing.T) {
		short := "hello"
		if got := Truncate(short, nil); got != short {
			t.Errorf("expected '%s', got '%s'", short, got)
		}
	})

	t.Run("longer than length", func(t *testing.T) {
		long := "hello world, this is a long string"
		result := Truncate(long, map[string]int{"length": 15})
		if len(result) > 15 {
			t.Errorf("expected <=15, got '%s' (%d)", result, len(result))
		}
	})

	t.Run("length 0", func(t *testing.T) {
		result := Truncate("hello", map[string]int{"length": 0})
		if result != "" {
			t.Errorf("expected '', got '%s'", result)
		}
	})

	t.Run("length < omissionLen", func(t *testing.T) {
		result := Truncate("hello world", map[string]int{"length": 1})
		if result != "" {
			t.Errorf("expected '', got '%s'", result)
		}
	})
}

func TestUnescape(t *testing.T) {
	if got := Unescape("a&amp;b&lt;c&gt;d&quot;e&#39;"); got != "a&b<c>d\"e'" {
		t.Errorf("expected 'a&b<c>d\"e'', got '%s'", got)
	}
	if got := Unescape("hello"); got != "hello" {
		t.Errorf("expected 'hello', got '%s'", got)
	}
}

func TestUpperCase(t *testing.T) {
	if got := UpperCase("--foo-bar--"); got != "FOO BAR" {
		t.Errorf("expected 'FOO BAR', got '%s'", got)
	}
}

func TestUpperFirst(t *testing.T) {
	if got := UpperFirst("fred"); got != "Fred" {
		t.Errorf("expected 'Fred', got '%s'", got)
	}
	if got := UpperFirst(""); got != "" {
		t.Errorf("expected '', got '%s'", got)
	}
}

func TestWords(t *testing.T) {
	result := Words("foo bar baz")
	if len(result) != 3 || result[0] != "foo" || result[1] != "bar" || result[2] != "baz" {
		t.Errorf("expected [foo bar baz], got %v", result)
	}
}

func TestTrim(t *testing.T) {
	if got := Trim("  abc  ", " "); got != "abc" {
		t.Errorf("expected 'abc', got '%s'", got)
	}
	if got := Trim("--abc--", "-"); got != "abc" {
		t.Errorf("expected 'abc', got '%s'", got)
	}
}

func TestTrimEnd(t *testing.T) {
	if got := TrimEnd("  abc  ", " "); got != "  abc" {
		t.Errorf("expected '  abc', got '%s'", got)
	}
}

func TestTrimStart(t *testing.T) {
	if got := TrimStart("  abc  ", " "); got != "abc  " {
		t.Errorf("expected 'abc  ', got '%s'", got)
	}
}
